### PR TITLE
Video optimization: use static list of supported types

### LINKED ISSUE #6012: Video Optimization: Define list of supported input file types
## Feature Description

For video transcoding we want to bail early if the input file is not supported by ffmpeg.

So if someone uploads an MKV file (video, not supported) we can reject right away, vs. if someone uploads a MOV file (supported), we can try transcoding it.

## Additional Context

With a command like `ffmpeg -demuxers` we can see which formats ffmpeg supports.

See also https://gist.github.com/DusanBrejka/35238dccb5cefcc804de1c5a218ee004

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

* In `useTranscodeVideo`, extract pats of `transcodeVideo` into a new function that simply returns an ffmpeg instance.
* Create new function `getSupportedInputFileTypes` (or similar) that loads ffmpeg, runs `ffmpeg -demuxers`

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

We want to improve detection of possibly supported video types.

## Summary

This PR replaces the simple `video/` check with a static list of supported input file types.

Of course it can't be guaranteed that the transcoding succeeds if the input file has a supported type, but this should improve detection before transcoding is even attempted.

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6012

### CODE REVIEW

File: assets/src/edit-story/constants/media.js [hunk 1]
```diff
@@ -17,3 +17,25 @@
 const GB_IN_BYTES = 1024 * 1024 * 1024;
 
 export const MEDIA_TRANSCODING_MAX_FILE_SIZE = 2 * GB_IN_BYTES;
+
+// List of mime types supported by the video transcoding.
+// Based on the demuxers list as of FFmpeg 4.3.
+export const MEDIA_TRANSCODING_SUPPORTED_INPUT_TYPES = [
```
Comment (spacedmonkey) [inline assets/src/edit-story/constants/media.js:23]: Should this be a variable in PHP with a filter like `get_allowed_image_mime_types`

https://github.com/google/web-stories-wp/blob/3438dca71e1b604111f70404e806207f9c7a21cf/includes/Traits/Types.php#L112-L140

So `get_allowed_transcodable_video_mime_types`

Comment (swissspidy) [reply assets/src/edit-story/constants/media.js:23]: What would be a good reason to offer such a filter?

This list is based on what ffmpeg actually supports. So one cannot simply _add_ more types to list, as it doesn't work / make sense.

And I don't see why someone would want to _remove_ something from the list.

Comment (spacedmonkey) [reply assets/src/edit-story/constants/media.js:23]: I agree, just checking to see if their was a reason for this.

**Comment by spacedmonkey** [comment]
Are there any tests we could add at this point? Just wondering as it seems like we don't have a lot of coverage at the moment.

**Comment by swissspidy** [comment]
> Are there any tests we could add at this point? Just wondering as it seems like we don't have a lot of coverage at the moment.

Good question.

We have 100% unit test coverage for the `useTranscodeVideo` and `useUploader` hooks.
Once we get to the snackbar/dialog notifications we can add integration tests for those.
Don't think e2e tests make sense here - don't want to do any actual video transcoding during tests.

Not sure what else we could do here in regards to tests.

**Comment by spacedmonkey** [comment]
Found this error while testing https://github.com/google/web-stories-wp/issues/6310

**Comment by spacedmonkey** [comment]
So what is meant to happen when I drag in an unsupported file?

Uploading [this](https://www.learningcontainer.com/mp4-sample-video-files-download/#Sample_MKV_File) mkv, I got no error message at all. 

Uploading [this](https://www.learningcontainer.com/mp4-sample-video-files-download/#Sample_MPG_File) mpg, I got blank error message at all. 

Screenshot 2021-02-10 at 16 27 48


I am not seeing this message at all, no matter what file I upload. 

https://github.com/google/web-stories-wp/blob/682909d912fd65a126cfd06644f0513aeb1ff116/assets/src/edit-story/app/uploader/useUploader.js#L112-L136

**Comment by swissspidy** [comment]
Looks like you've uncovered some flaws in the pre-existing uploading logic. These are not necessarily related to this PR or the video optimization in general...

First, lets look at these two files and their types:

```bash
$ file --mime-type sample-mkv-file.mkv 
> sample-mkv-file.mkv: video/x-matroska

$ file --mime-type sample-mpg-file.mpg 
> sample-mpg-file.mpg: video/mpeg
```

So far, so good! Both types are on the list here, and ffmpeg transcodes them just fine.

Let's try to upload the files!

**Uploading `sample-mkv-file.mkv`**

Using the debugger I noticed that Chrome does not support this file type at all. `getResourceFromLocalFile` returns `null` for because `file.type` is an empty string instead of the expected `video/x-matroska`.

That's just the browser outright not recognizing the file. We cannot really upload it this way (we don't know its type), so we have to drop it. And that's what `useUploadMedia` does - it drops the file:

https://github.com/google/web-stories-wp/blob/1ea476bb315d2618cfdd95f6ffd9775593064a69/assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js#L65

The problem: it drops it silently.

There would need to be a check for this case and display a snackbar message like "Invalid file type" or something.

I just did this here real quick.

**Uploading `sample-mpg-file.mpg`**

Here Chrome _does_ recognize the `video/mpeg` file type in `getResourceFromLocalFile` and tries to get a resource. However, this fails and causes an error.

Basically what you are experiencing in #6310, but in this case there is no error message. `e.message` does simply not exist.

https://github.com/google/web-stories-wp/blob/1ea476bb315d2618cfdd95f6ffd9775593064a69/assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js#L80-L88

The solution here is to provide a fallback message so that at least _something_ is being displayed.

I just did this here real quick.

**Comment by swissspidy** [comment]
> > Are there any tests we could add at this point? Just wondering as it seems like we don't have a lot of coverage at the moment.
> 
> Good question.
> 
> We have 100% unit test coverage for the `useTranscodeVideo` and `useUploader` hooks.
> Once we get to the snackbar/dialog notifications we can add integration tests for those.
> Don't think e2e tests make sense here - don't want to do any actual video transcoding during tests.
> 
> Not sure what else we could do here in regards to tests.

I'll update the `useUploadMedia` unit tests to cover the cases we just encountered here.

**Comment by spacedmonkey** [comment]
I am now not seeing any files transcode at all. I either get these two messages. On anything other an mp4/webm. 

Screenshot 2021-02-11 at 11 23 02
Screenshot 2021-02-11 at 11 22 40

Am I missing something?

**Comment by swissspidy** [comment]
> I am now not seeing any files transcode at all. I either get these two messages. On anything other an mp4/webm.
> 
> Screenshot 2021-02-11 at 11 23 02
> 
> Screenshot 2021-02-11 at 11 22 40
> 
> Am I missing something?

No, you're not. There's indeed a bug. But this is not isolated to this PR. It also happens on `main`. So it shouldn't block review/merge.

We should fix this as part of #6310. It's not an issue with the transcoding, but how the editor tries to display the video in the library  while it's uploading.

**Review by spacedmonkey — APPROVED** [review]
Sorry for blocking this issue, didn't know that is was broken in main. Merging as this PR makes things better.
